auth_partner/models/auth_partner.py

```diff
@@ -0,0 +1,307 @@
+                    "password. "
+                )
+            )
+
+        if replacement is not None:
+            auth_partner.encrypted_password = replacement
+
+        return auth_partner
+
+    @api.model
+    def _validate_email(self, directory, token):
+        auth_partner = directory._decode_token(token, "validate_email")
+        auth_partner.write({"mail_verified": True})
+        return auth_partner
+
+    def _get_impersonate_url(self, token, **kwargs):
+        # You should override this method according to the impersonation url
+        # your framework is using
+
+        base = self.env["ir.config_parameter"].sudo().get_param("web.base.url")
+        url = f"{base}/auth/impersonate/{token}"
+        return url
+
+    def _get_impersonate_action(self, token, **kwargs):
+        return {
+            "type": "ir.actions.act_url",
+            "url": self._get_impersonate_url(token, **kwargs),
+            "target": "new",
+        }
+
+    def impersonate(self):
+        self.ensure_one()
+        if self.env.user not in self.impersonating_user_ids:
+            raise AccessDenied(_("You are not allowed to impersonate this user"))
+
+        token = self._generate_impersonating_token()
+        return self._get_impersonate_action(token)
+
+    @api.depends_context("uid")
+    def _compute_user_can_impersonate(self):
+        for record in self:
+            record.user_can_impersonate = self.env.user in record.impersonating_user_ids
+
+    @api.model
+    def _impersonating(self, directory, token):
+        partner_auth = directory._decode_token(
+            token,
+            "impersonating",
+            key_salt=lambda auth_partner: (
+                auth_partner.date_last_impersonation.isoformat()
+                if auth_partner.date_last_impersonation
+                else "never"
+            ),
+        )
+        partner_auth.date_last_impersonation = fields.Datetime.now()
+        return partner_auth
+
+    def _on_reset_password_sent(self):
+        self.ensure_one()
+        self.date_last_request_reset_pwd = fields.Datetime.now()
+        self.date_last_sucessfull_reset_pwd = None
+        self.nbr_pending_reset_sent += 1
+
+    def _send_invite(self):
+        self.ensure_one()
+        self.directory_id._send_mail_background(
+            "set_password",
+            self,
+            callback_job=self.delayable()._on_reset_password_sent(),
+            token=self._generate_set_password_token(),
+        )
+
+    def send_invite(self):
+        for rec in self:
+            rec._send_invite()
+
+    def _request_reset_password(self):
+        return self.directory_id._send_mail_background(
+            "reset_password",
+            self,
+            callback_job=self.delayable()._on_reset_password_sent(),
+            token=self._generate_set_password_token(),
+        )
+
+    def _set_password(self, directory, token, password):
```

Comment:
Can you add some comment here, (or in the decode_token method ?) regarding the key_salt ?
I know that you add this logic to make the token usable only one (as the function depend on the encrypted password)
But new contributor need help to understand this logic ?

Thanks